File: ios/Packages/GemstoneServices/Sources/Stores/PriceStore.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import typealias Gemstone.AssetId
import struct Gemstone.AssetMarket
import struct Gemstone.AssetPrice
import enum Gemstone.Currency
import struct Gemstone.FiatRate
import protocol Gemstone.GemPriceStore
import struct Gemstone.GemPriceUpdate
import GemstonePrimitives
import Primitives
import Store

public final class GemstonePriceStore: GemPriceStore, @unchecked Sendable {
    private let priceStore: PriceStore

    public init(priceStore: PriceStore) {
        self.priceStore = priceStore
    }

    public func getPrices(assetIds: [Gemstone.AssetId]) throws -> [Gemstone.AssetPrice] {
        try priceStore.getPrices(for: assetIds).map { $0.toGem() }
    }

    public func getRate(currency: Gemstone.Currency) async throws -> Gemstone.FiatRate? {
        let currency = currency.toPrimitives()
        return try priceStore.getRate(currency: currency.rawValue).map { Primitives.FiatRate(symbol: currency, rate: $0.rate).toGem() }
    }

    public func getRates() async throws -> [Gemstone.FiatRate] {
        try priceStore.getRates().map { Primitives.FiatRate(symbol: $0.symbol, rate: $0.rate).toGem() }
    }

    public func saveRatesAndPrices(currency _: Gemstone.Currency, rates: [Gemstone.FiatRate], conversion: Gemstone.FiatRate?, prices: [GemPriceUpdate]) async throws {
        try priceStore.saveRatesAndPrices(rates.map { $0.toPrimitives() }, conversion: conversion?.toPrimitives(), prices: prices.map { try $0.priceUpdate() })
    }

    public func savePrices(currency _: Gemstone.Currency, prices: [GemPriceUpdate]) async throws {
        try priceStore.updatePrices(prices.map { try $0.priceUpdate() })
    }

    public func convertPrices(currency _: Gemstone.Currency, rate: Double) async throws {
        try priceStore.convertPrices(rate: rate)
    }

    public func saveMarket(assetId: Gemstone.AssetId, market: Gemstone.AssetMarket) async throws {
        try priceStore.updateMarket(assetId: Primitives.AssetId(id: assetId), market: market.toPrimitives())
    }
}

private extension GemPriceUpdate {
    func priceUpdate() throws -> PriceUpdate {
        try PriceUpdate(
            assetId: Primitives.AssetId(id: assetId),
            price: price,
            priceUsd: priceUsd,
            priceChangePercentage24h: priceChangePercentage24h,
            updatedAt: updatedAt,
        )
    }
}

File: android/data/services/gemstone/src/main/kotlin/com/gemwallet/android/data/services/gemstone/stores/PriceStore.kt
package com.gemwallet.android.data.services.gemstone.stores

import com.gemwallet.android.data.services.store.database.AssetsDao
import com.gemwallet.android.data.services.store.database.PricesDao
import com.gemwallet.android.data.services.store.database.entities.DbPrice
import com.gemwallet.android.data.services.store.database.entities.toDTO
import com.gemwallet.android.data.services.store.database.entities.toRecord
import com.gemwallet.android.ext.toGem
import com.gemwallet.android.ext.toIdentifier
import com.gemwallet.android.ext.toPrimitives
import com.wallet.core.primitives.AssetId
import com.wallet.core.primitives.Currency
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.firstOrNull
import uniffi.gemstone.AssetPrice
import uniffi.gemstone.GemPriceStore
import uniffi.gemstone.GemPriceUpdate

class GemstonePriceStore(private val pricesDao: PricesDao, private val assetsDao: AssetsDao) : GemPriceStore {

    override suspend fun getPrices(assetIds: List<String>): List<AssetPrice> = pricesDao.getByAssets(assetIds).map { it.toAssetPrice() }

    override suspend fun getRate(currency: uniffi.gemstone.Currency): uniffi.gemstone.FiatRate? = pricesDao.getRates(currency.toPrimitives()).firstOrNull()?.toDTO()?.toGem()

    override suspend fun getRates(): List<uniffi.gemstone.FiatRate> = pricesDao.getRates().toDTO().map { it.toGem() }

    override suspend fun saveRatesAndPrices(currency: uniffi.gemstone.Currency, rates: List<uniffi.gemstone.FiatRate>, conversion: uniffi.gemstone.FiatRate?, prices: List<GemPriceUpdate>) = pricesDao.saveRatesAndPrices(
        rates = rates.map { it.toPrimitives().toRecord() },
        conversion = conversion?.toPrimitives()?.toRecord(),
        prices = prices.map { it.toRecord(currency.toPrimitives()) },
    )

    override suspend fun savePrices(currency: uniffi.gemstone.Currency, prices: List<GemPriceUpdate>) {
        pricesDao.insert(prices.map { it.toRecord(currency.toPrimitives()) })
    }

    override suspend fun convertPrices(currency: uniffi.gemstone.Currency, rate: Double) = pricesDao.updateValues(currency.toPrimitives(), rate)

    override suspend fun saveMarket(assetId: String, market: uniffi.gemstone.AssetMarket) {
        assetsDao.setMarket(market.toPrimitives().toRecord(AssetId(assetId)))
    }

    fun observeUsdPrice(assetId: AssetId): Flow<Double?> = pricesDao.getUsdPrice(assetId.toIdentifier())
}

private fun GemPriceUpdate.toRecord(currency: Currency) = DbPrice(
    assetId = assetId,
    value = price,
    usdValue = priceUsd,
    dayChanged = priceChangePercentage24h,
    currency = currency,
    updatedAt = updatedAt,
)
